File: leetcode/src/test/java/com/giwankim/leetcode/RemoveNthFromEndTest.java
package com.giwankim.leetcode;

import static org.assertj.core.api.Assertions.assertThat;

import com.giwankim.leetcode.support.ListNode;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class RemoveNthFromEndTest {
  private RemoveNthFromEnd sut;

  @BeforeEach
  void setUp() {
    sut = new RemoveNthFromEnd();
  }

  @Test
  void removeNthFromEnd() {
    ListNode head = ListNode.of(1, 2, 3, 4, 5);
    ListNode expected = ListNode.of(1, 2, 3, 5);

    ListNode actual = sut.removeNthFromEnd(head, 2);

    assertThat(actual).isEqualTo(expected);
  }

  @Test
  void removeLastNode() {
    ListNode head = ListNode.of(1, 2);
    ListNode expected = ListNode.of(1);

    ListNode actual = sut.removeNthFromEnd(head, 1);

    assertThat(actual).isEqualTo(expected);
  }

  @Test
  void removeOneResultsInEmptyList() {
    ListNode head = ListNode.of(1);
    ListNode expected = ListNode.of();

    ListNode actual = sut.removeNthFromEnd(head, 1);

    assertThat(actual).isEqualTo(expected);
  }
}

File: leetcode-kt/src/test/kotlin/com/giwankim/leetcode/RemoveNthFromEndTest.kt
package com.giwankim.leetcode

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe

class RemoveNthFromEndTest :
    FunSpec(
        {
            lateinit var sut: RemoveNthFromEnd

            beforeTest { sut = RemoveNthFromEnd() }

            test("remove last node, then empty list") {
                val head = ListNode.of(1)
                val expected = ListNode.of()
                sut.removeNthFromEnd(head, 1) shouldBe expected
            }

            test("remove first node from length 2 list") {
                val head = ListNode.of(1, 2)
                val expected = ListNode.of(1)
                sut.removeNthFromEnd(head, 1) shouldBe expected
            }

            test("remove second to last node") {
                val head = ListNode.of(1, 2, 3, 4, 5)
                val expected = ListNode.of(1, 2, 3, 5)
                sut.removeNthFromEnd(head, 2) shouldBe expected
            }
        },
    )
